### PR TITLE
30-Squad Implementation of HistoryPage(Epic 6.4)

### PR DESCRIPTION
**Pull request by pleroux64**
# Description
30-Squad has implemented the History Page feature to the Kai Website. This feature adds a new button in the Nav menu that a user can click on to access the history page. This history page contains cards that correspond to instances of when user used a tool like the flashcard or MC generator. These cards are sorted by date, and have information displayed on which tool was used, and how it was used. These cards also have a preview button, which brings up a slide panel with the actual tool data generated from the tool instance card they clicked.

Shoutout to Reza aka [Cosmodocus](https://github.com/Cosmodocus), team lead of Squad 20, for letting us use their listener-oriented redux thunk/slice code.
Borrowed Code: frontend/redux/thunk/toolHistory.js, frontend/redux/slices/toolHistorySlice.

Domain: 
[test4-2a3ef.web.app](https://test4-2a3ef.web.app/)

LOOM links: 
Part 1: 
https://www.loom.com/share/c360998ae07541d9a429995d5c18b3ff?sid=f5641246-e059-490b-8d6b-fe25f9c2deed

Part 2: 
https://www.loom.com/share/963680aec56344a48ec3e56d2b396e8c?sid=19040d68-eb27-49a0-904b-8a95f392a00a

Part 3:
https://www.loom.com/share/11cf0941430c4330a0f5b7f2fb221700?sid=277fc0cb-45d3-4a90-a72d-36e7fc3aaaf8
8FDD516B-CE65-4C55-8C06-66EDF3166664

MCQ variant: 
D702F48E-B431-4250-881A-ECA3EE7DEBFB

Ticket:

## Type of change
Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Include screenshot/video if possible.
- [ ] Test A - Use the tools and see if the history page generates a card with all the same information from the tool use.
- [ ] Test B - Use the firestore to assign different dates, and verify the history page sorts by date.
- [ ] Test C - Use the preview button on the Card, to see if a slide panel with all the tool data is displayed with the respective tool 
- [ ] Test D - Use the copy button, and paste it somewhere to see the flashcard or MCQ data.
- [ ] Test E - Use the export button and look at the file to see a table with the tool data, should be different for MCQ and FlashCards

### CODE REVIEW

**Comment by bkb-Git** [comment]
@pleroux64 Looks mostly good, dropped a few comments that would require changes to be made in the codebase and I have a few questions concerning certain aspects of the UI which I'll get to during today's dev workins session.

File: frontend/redux/slices/toolsSlice.js
Comment (pleroux64) [inline frontend/redux/slices/toolsSlice.js:1]: test

**Review by pleroux64 — COMMENTED** [review]
test

**Review by bkb-Git — CHANGES_REQUESTED** [review]
Please look at the comments and make the requested changes.

File: frontend/services/history/transformToolData.jsx [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+      } else {
+        notableConcepts = primaryConcept;
+      }
+
+      title = `${toolDetails.titlePrefix} ${primaryConcept} and More`;
+      content = `${toolDetails.descriptionPrefix} ${notableConcepts}`;
+    } else {
+      title = `${toolDetails.titlePrefix} ${topic}`;
+      content = `${toolDetails.descriptionPrefix} ${topic}`;
+    }
```
Comment (bkb-Git) [inline frontend/services/history/transformToolData.jsx:41]: Let's abstract this as a default `header || setDefault()`

File: frontend/utils/DateUtils.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
+import moment from 'moment';
+
+function categorizeDate(creationDate, currentDate) {
```
Comment (bkb-Git) [inline frontend/utils/DateUtils.js:3]: Let's stick to a single pattern on how we are creating functions, in this case we use `const categorizeData = () => {}`

**Review by bkb-Git — CHANGES_REQUESTED** [review]
Left a few comments that need to be resolved

**Comment by pleroux64** [comment]
@bkb-Git 
Made some changes to make the drawer more dynamic, fixed the redux and added the snapshot, and some other small things

File: functions/controllers/kaiAIController.js [hunk 3]
```diff
@@ -264,6 +274,26 @@ app.post('/api/tool/', (req, res) => {
...
+  try {
+    const toolSessionRef = await admin
+      .firestore()
+      .collection('toolSessions')
+      .add({
+        ...sessionData,
+        createdAt: Timestamp.fromMillis(Date.now()),
+      });
+
+    logger.log(`Tool session saved with ID: ${toolSessionRef.id}`);
```
Comment (bkb-Git) [inline functions/controllers/kaiAIController.js:292]: Let's set a flag called DEBUG in order to display this log when we are debugging, don't need this to log on every instance

File: frontend/services/toolHistory/transformToolData.jsx
Comment (bkb-Git) [inline frontend/services/toolHistory/transformToolData.jsx:1]: This file shouldn't be under services, it's a util

**Review by bkb-Git — CHANGES_REQUESTED** [review]
Please take a look at some of the comments and make the necessary changes.

**Pull request merged**